### PR TITLE
First iteration of the Database API

### PR DESCRIPTION
**Pull request by Jkutkut**
First iteration of the DB API.

Note: The issue #19 will be kept open to continue working on new features.

### CODE REVIEW

**Review by aleexgheorghe — APPROVED** [review]
The implemented methods seem to be fine! I will add more methods to the issue as I need them 🎉.

File: app/CineHub/app/src/main/java/org/cinehub/api/result/OnFailureCallback.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package org.cinehub.api.result;
+
+/**
+ * Callback to be called on failure.
+ * @param <T> type of the value to be passed to the callback
+ * @author Jkutkut
+ */
+public interface OnFailureCallback<T> {
```
Comment (alexaib2002) [inline app/CineHub/app/src/main/java/org/cinehub/api/result/OnFailureCallback.java:9]: As this interface is meant to be used as a lambda expression, it should have a `@FunctionalInterface` annotation on top of the declaration

File: app/CineHub/app/src/main/java/org/cinehub/api/result/OnSuccessCallback.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package org.cinehub.api.result;
+
+/**
+ * Callback to be called on success when no value is returned.
+ *
+ * @author Jkutkut
+ */
+public interface OnSuccessCallback {
```
Comment (alexaib2002) [inline app/CineHub/app/src/main/java/org/cinehub/api/result/OnSuccessCallback.java:9]: Same as OnFailureCallback, it should have the `@FunctionalInterface` annotation

File: app/CineHub/app/src/main/java/org/cinehub/api/result/OnSuccessValueCallback.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package org.cinehub.api.result;
+
+/**
+ * Listener to be called on success.
+ * @param <T> type of the value to be returned
+ * @author Jkutkut
+ */
+public interface OnSuccessValueCallback<T> {
```
Comment (alexaib2002) [inline app/CineHub/app/src/main/java/org/cinehub/api/result/OnSuccessValueCallback.java:9]: Same as OnFailureCallback, it should have the @FunctionalInterface annotation

**Review by alexaib2002 — CHANGES_REQUESTED** [review]
Everything looks very promising, but interfaces with one method intended to be used as lambda expressions should have the `@FunctionalInterface` annotation on top of the interface declaration!

Also, I would like to point out that making changes on `MainActivity` doesn't seem necessary, as this is an API, all testing could be done with the `ExampleUnitTest` class. Making changes on a deleted file of another branches may end up on merging conflicts. Still, I will approve this, as `MainActivity` can be deleted on the merging process.

Summarizing, I think the only things that must be changed are the `FunctionalInterface` ones. Good work!

**Comment by alexaib2002** [comment]
@Jkutkut, on second thought I think the Callback interfaces may not even be necessary: Both `OnFailureCallback` and `OnSuccessValueCallback` can be replaced with the builtin `Callable<T>` Java interface. As for the `OnSuccessCallback`, it could be replaced with `Runnable`.

This would simplify the API class structure, but code may be more difficult to read on the long run. The choice is up to you!

**Comment by Jkutkut** [comment]
@alexaib2002 
- @FunctionalInterface: Sure, it will be implemented now.
- ExampleUnitTest: I strongly agree. Furthermore, I was already working on it. However, I met an error: Firebase requests the context of the current activity to work. I am working on a way to test these methods using `InstrumentalTests`, where an applicaction context is generated and the test ir run on the device. However, due to the urgency of deploying a first iteration of the API, I prioritized the logic using the main activity as a tool for testing. This should not be a problem anyway because this activity does not exists in the UI/UX development branches (and will be removed, as you mentioned).
- Runnable and Callable: It was one of the options I had but I found the same issue as you mentioned: readability. For this reason, I will continue using this custom interfaces.

**Review by alexaib2002 — APPROVED** [review]
All issues solved, changes approved. Unit testing can be implemented on future revisions of #19

**Pull request merged**